src/Features/Core/Portable/CodeRefactorings/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,157 @@
+            {
+                if (condition == null)
+                {
+                    switchDefaultBody = body;
+                    break;
+                }
+
+                var operands = GetLogicalOrExpressionOperands(condition);
+                var patterns = new List<TPattern>();
+
+                // Iterate over "||" operands to make a case label per each condition.
+                foreach (var operand in operands.Reverse())
+                {
+                    var pattern = CreatePatternFromExpression(operand, semanticModel, ref switchExpression);
+                    if (pattern == null)
+                    {
+                        return;
+                    }
+
+                    patterns.Add(pattern);
+                }
+
+                switchSections.Add((patterns, body));
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c =>
+                UpdateDocumentAsync(root, document, ifStatement, switchDefaultBody, switchExpression, semanticModel, switchSections)));
+        }
+
+        protected bool AreEquivalent(TExpressionSyntax expression, ref TExpressionSyntax switchExpression)
+        {
+            // If we have not figured the switch expression yet,
+            // we will assume that the first expression is the one.
+            if (switchExpression == null)
+            {
+                switchExpression = expression;
+                return true;
+            }
+
+            return AreEquivalentCore(expression, switchExpression);
+        }
+
+        private static bool IsConstant(TExpressionSyntax node, SemanticModel semanticModel)
+            => semanticModel.GetConstantValue(node).HasValue;
+
+        protected static bool TryDetermineConstant(
+            TExpressionSyntax expression1,
+            TExpressionSyntax expression2,
+            SemanticModel semanticModel,
+            out TExpressionSyntax constant,
+            out TExpressionSyntax expression)
+        {
+            (constant, expression) =
+                    IsConstant(expression1, semanticModel) ? (expression1, expression2) :
+                    IsConstant(expression2, semanticModel) ? (expression2, expression1) :
+                    default((TExpressionSyntax, TExpressionSyntax));
+
+            return constant != null;
+        }
+
+        protected abstract SyntaxToken GetIfKeyword(TIfStatementSyntax ifStatement);
+        protected abstract bool AreEquivalentCore(TExpressionSyntax expression, TExpressionSyntax switchExpression);
+        protected abstract bool CanConvertIfToSwitch(TIfStatementSyntax ifStatement, SemanticModel semanticModel);
+        protected abstract IEnumerable<TExpressionSyntax>
+            GetLogicalOrExpressionOperands(TExpressionSyntax syntaxNode);
+
+        protected abstract IEnumerable<(TStatementSyntax, TExpressionSyntax)>
+            GetIfElseStatementChain(TIfStatementSyntax ifStatement);
```

Comment:
C# doesn't have trailing return types so in case if long types involving tuples wrapping after method name won't really help. I'll change it regardless.